rfc/20241021-dynamic-provider-instances.md

```diff
@@ -0,0 +1,766 @@
+
+# Allowing `for_each` without `alias`
+
+The main differences in this proposal vs. the original early eval proposal are in the treatment of instance keys: this proposal calls for them to be added _in addition to_ aliases rather than as a different way of specifying aliases, and then further proposes that `for_each` only be allowed in conjunction with `alias`.
+
+The second part of that is not techncially required for correctness and is instead motivated by user experience concerns. Consider the following hypothetical configuration:
+
+```hcl
+provider "aws" {
+  alias = "a"
+
+  # ...
+}
+
+provider "aws" {
+  for_each = {
+    "a" = {}
+  }
+
+  # ...
+}
+```
+
+If we allow this then we need some syntax for distinguishing between these two cases. We _do_ have the syntax tools required for this, and existing conventions would perhaps call for the first block's singleton instance to be called `aws.a` while the second's instance is called `aws["a"]`. 
+
+OpenTofu would have no technical problem with those two addresses being distinct, but there is no other place in the OpenTofu language where `.a` and `["a"]` are both allowed but yet have different meanings.
+
+Requiring `alias` to always be set when `for_each` is set is a relatively-small concession that means that we have only three forms to deal with and each builds on the previous in a similar way to how other similar language features work:
+
+* `aws`: default provider configuration
+* `aws.secondary`: singleton additional provider configuration
+* `aws.by_region["us-east-2"]`: multi-instance additional provider configuration
+
+Although it's a much less significant concern than user confusion, this constraint also allows for simpler address decoding logic: the number of traversal steps immediately distinguishes between the three cases, and each position has only one valid syntax element: the sequence is always root name, attribute access, index. Simpler decoding logic typically also allows better error messages for invalid input, because the error message doesn't need to hedge so much between multiple possible user intents.
+
+Since the concept of "default provider configurations" exists primarily to allow OpenTofu to magically select or even generate them when there's no explicit selection, and that behavior is only possible if the default configuration is a singleton, I think it's justified to require a static alias so that there are fewer situations to explain in the documentation and no risk of conflict with existing assumptions about "default provider configurations".
```

Comment:
I think it's also worth noting in this section that by requiring provider aliases, it would preserve the capability mentioned above about accessing resources/provider by name/alias as if they were a map.  `aws_bucket["foo"]` vs `aws_bucket.foo` for example